### PR TITLE
OCPBUGS-33913, OCPBUGS-34261: CurrentImagePullSecret should be consumed by the MCD

### PR DESCRIPTION
**Pull request by cheesesashimi**
**- What I did**

The CustomImagePullSecret that is referred to by a MachineOSConfig was not being used. This means that pulling an OS image from a private image registry will not work. This work does the following:

1. Migrates functions for changing the formatting of pull secrets from the BuildController package into controller/common, allowing reuse.
2. Adds secret volume mounts to the MachineConfigDaemon DaemonSet as well as template rendering logic to collect that information whenever a MachineOSConfig is changed.
3. Adds an MCD-side merge process that reads the secrets from the volume mounts, reads the secrets provided by the ControllerConfig, and then merges them into /etc/mco/internal-image-pull-secret.json.
4. Adds unit and e2e tests to validate that a secret referenced by a MachineOSConfig is written to the nodes' filesystem, overriding any values from the ControllerConfig.

**- How to verify it**

1. Bring up a cluster with the TechPreviewNoUpgrade feature-gate added.
2. Set up a MachineOSConfig to use a private image registry for both pushing and pulling.
3. Create a MachineConfigPool and a MachineOSConfig and wait for the build to complete. While this is occurring the MCD DaemonSet will restart since it will now have a secret mounted into it. If multiple MachineOSConfigs are present, multiple secret volumes will be mounted into the MCD pod, even if they are the same secret.
4. Move a node into the new MachineConfigPool and wait for it to drain and reboot.
5. When the node comes back up, it should be booted into the OS image that was pulled from the private image registry.

**- Description for the changelog**
CurrentImagePullSecret should be used by MachineConfigDaemon

### CODE REVIEW

**Comment by cheesesashimi** [comment]
@sinnykumari That is correct. I used that PR as the base for this one because of the overlap.

**Review by yuqi-zhang — APPROVED** [review]
Haven't tested by hand but the logic itself looks fine to me. I'm in favour of attaching both bugs and closing https://github.com/openshift/machine-config-operator/pull/4372 since this is fully rebased and tested

Given that this is a blocker bug, will leave the

/lgtm
/hold

And let @sergiordlr remove the hold once pre-merge testing has passed. Thanks!

**Comment by sergiordlr** [comment]
Verification of the internal-regsitry-pull-secret file

Passed:
`"[sig-mco] MCO Layering Author:sregidor-ConnectedOnly-Longduration-NonPreRelease-Medium-54056-Update osImage using the internal registry to store the image [Disruptive] [Serial]"`


The file is not present in the rendered machine-configs
```
$ oc get mc -oyaml rendered-worker-de597a802ef5cd27f6c53083f93a89c3 | grep "path:" | grep internal
$ oc get mc -oyaml rendered-master-13a425be563ebc405e0550d40f658ee3  | grep "path:" | grep internal

```

The file content was the same content stored in .spec.internalRegistryPullSecret in the controllerconfig resource.


No issues found.

**Comment by cheesesashimi** [comment]
Couple of things:

> When we create a MOSC without buildOutputs secret mchine-config CO is degrade with the wrong message

I feel like this shouldn't be possible given API validation, though evidently, it is possible. To prevent this, we can check for the presence of all of the secrets referenced by the MachineOSConfig and emit an error if an expected secret is not present. I feel like such a thing would also be covered by API validation, but I don't know for sure.

> Events seem to show a wrong message, bad format

That feels like a regression though I don't think it was caused by this PR. I will take a look though since the fix should be pretty simple.

> When 2 machineosbuilds are created, and we remove the MSOC, not all machineosbuilds are garbage-collected

I don't think we have a good story around garbage collection right now. That said, I suspect the root cause here is that the `machine-os-builder` pod might be shutting down before the second deletion can take place. I don't think that was introduced in this PR, however.

> MCP are stuck doing nothing after the OCL image is built. The time is more or less random, it can take up to 27 minutes to start applying the new image.

Question: When the MCP is stuck in this state, what are the MCDs doing? Specifically, what I'm asking is: Is the DaemonSet rolling out a new MCD? Or are they all up and running? Additionally, do they all have the secret volume mounts for the secrets present in the MachineOSConfigs? This will help clarify whether this PR is what is causing this or if the root cause lies elsewhere.

**Comment by cheesesashimi** [comment]
For posterity, I looked at the must-gather provided by @sergiordlr and also looked at the cluster he was running everything on. I observed a few interesting things, but cannot definitively determine whether this PR introduces this rollout delay. Here's why:

- When the MCO updates the DaemonSet object for the MachineConfigDaemon, all of the pods restart in order to pick up the new secrets. This occurs while the build is running, but a subset of the MCD pods fail the health / liveness check. I was unable to see logs indicating what was happening inside those MCD pods in order to determine whether it was related to the changes this PR makes.
- The NodeController does not begin to set annotations on the nodes to roll out the new config for quite some time after the build has completed. Meanwhile, just updating the MCD DaemonSet object should not cause the NodeController to not select a given node for updates.

Theoretically, what could happen is that the MCD pod could fail to start if the secret it is configured to use does not actually exist (unless one sets `optional: true` on the volume config, which we are not doing). Indeed, I saw that specific scenario a few times while I was trying to write the e2e test for this. In that scenario, replacement MCD pods would be blocked from rolling out given the default rollout strategy we're using for them. Similarly, if one of the MCD pods is configured to block SIGTERMS (like it does whenever an update is in process), that could block the rollout of new MCD pods since the DaemonSet controller will not be able to terminate that pod via a SIGTERM.

The following solutions exist for both of those scenarios:
- I've added a validator to that path to ensure that those secrets exist before modifying the DaemonSet object.
- We can update the rollout strategy to tolerate more than one MCD pod being unavailable at a given time. I'd like to add that only if we've validated that we've run into issues with SIGTERMs.

**Comment by cheesesashimi** [comment]
Update: I was able to reproduce the MCD health / liveness check in my sandbox cluster. While this occurred, I was able to stream the pod logs using k9s. I observed that the reason why the checks were temporarily failing is because we are starting up the health endpoint fairly late in the MCD startup process. At this point, I can safely rule that out as being the root cause of the delayed image rollout once the image is built.

**Comment by sergiordlr** [comment]
Verified using IPI on GCP

- Now the  secrets are reporting a descriptive error when they are not properly configured

- Now the events are displaying a properly formatted message 

```
$ oc get events |grep SetDesiredConfig
75m         Normal    SetDesiredConfigAndOSImage                 machineconfigpool/worker                                          Targeted node sregidor-voclsec6-knd2c-worker-a-m2krj to MachineConfig: rendered-worker-269adf05864ad9d63d7954781af64f0b / Image: image-registry.openshift-image-registry.svc:5000/mco-tc-73947/ocb-image@sha256:1349bbff7e3e4a2e3314ae98af7b96a968c0aa782adee746998bf828e9204f13

$ oc get events |grep SetDesiredConfigAndOSImage 
75m         Normal    SetDesiredConfigAndOSImage                 machineconfigpool/worker                                          Targeted node sregidor-voclsec6-knd2c-worker-a-m2krj to MachineConfig: rendered-worker-269adf05864ad9d63d7954781af64f0b / Image: image-registry.openshift-image-registry.svc:5000/mco-tc-73947/ocb-image@sha256:1349bbff7e3e4a2e3314ae98af7b96a968c0aa782adee746998bf828e9204f13
```

- Regarding the time gap between MCO builds the image and MCO starts applying the image, we have created a new jira ticket to track this issue https://issues.redhat.com/browse/OCPBUGS-35509


- Regarding the machineosbuild resources that are not garbage collected, we have created a new jira ticket to track this behaviour https://issues.redhat.com/browse/OCPBUGS-35512


We have verified that with this fix OCL functionality can properly use the buildOutputs.currentImagePullSecret secret. Hence, we add the qe-approved label. The issues  mentioned in this PR will be fixed and tracked using the jira tickets that we have created for that.

/label qe-approved

**Comment by cheesesashimi** [comment]
Regarding the update latency bug ([OCPBUGS-35509](https://issues.redhat.com/browse/OCPBUGS-35509)): This PR does not introduce that problem. Instead, it is caused by the lack of informers within NodeController for both the MachineOSConfig and MachineOSBuild objects. See bug for further discussion.

File: pkg/daemon/certificate_writer.go
Comment (yuqi-zhang) [inline pkg/daemon/certificate_writer.go:480]: So if I am understanding this correctly, this shouldn't exist unless you have OS builds turned on and the secret is mounted to the daemonset?

According to the logs (e.g. https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_machine-config-operator/4395/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1801305805287329792/artifacts/e2e-gcp-op/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-daemon-cc462_machine-config-daemon.log) this seems to be the case (which I think is normal). We can probably also turn down the verbosity here

File: pkg/operator/sync.go
Comment (yuqi-zhang) [inline pkg/operator/sync.go:616]: Hmm, does this need to be behind a FG?

**Review by yuqi-zhang — COMMENTED** [review]
Some nits inline, also another question, the controller complains that it can't watch MachineOSBuild objects (e.g. https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_machine-config-operator/4395/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1801305805287329792/artifacts/e2e-gcp-op/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-controller-6458cfc9c-t7l9x_machine-config-controller.log).

I don't think that's being caused by this PR and I remember we had something to track that, could you remind me where that is if there is a bug for it?

File: pkg/daemon/certificate_writer.go
Comment (cheesesashimi) [reply pkg/daemon/certificate_writer.go:480]: Agreed. I'll put it behind a `V(4)`.

File: pkg/daemon/certificate_writer.go
Comment (cheesesashimi) [reply pkg/daemon/certificate_writer.go:383]: Agreed. I'll put it behind a `V(4)`.

File: pkg/operator/sync.go
Comment (cheesesashimi) [reply pkg/operator/sync.go:616]: It may not look like it, but it technically is. The `optr.getAndValidateMachineOSConfigs()` function checks for the presence of the FeatureGate before trying to perform its query. If the FeatureGate is not enabled, it returns nil. I can add some comments or the like to indicate that is the case.

File: pkg/daemon/certificate_writer.go
Comment (cheesesashimi) [reply pkg/daemon/certificate_writer.go:480]: That is correct. If the root path here does not exist, we don't go any further with respect to the secret reconciliation.

**Pull request merged**
